ccl/changefeedccl: TestAvroEncoder failed under stress

SHA: https://github.com/cockroachdb/cockroach/commits/965525a5deb3c4fab5ab232e069846ecc08f632d

Parameters:
```
TAGS=
GOFLAGS=-race
```

To repro, try:

```
# Don't forget to check out a clean suitable branch and experiment with the
# stress invocation until the desired results present themselves. For example,
# using stress instead of stressrace and passing the '-p' stressflag which
# controls concurrency.
./scripts/gceworker.sh start && ./scripts/gceworker.sh mosh
cd ~/go/src/github.com/cockroachdb/cockroach && \
stdbuf -oL -eL \
make stressrace TESTS=TestAvroEncoder PKG=github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl TESTTIMEOUT=5m STRESSFLAGS='-maxtime 20m -timeout 10m' 2>&1 | tee /tmp/stress.log
```

Failed test: https://teamcity.cockroachdb.com/viewLog.html?buildId=1133737&tab=buildLog

```
I190212 07:34:23.816867 39761 sql/event_log.go:135  [n1,intExec=create-default-db] Event: "create_database", target: 51, info: {DatabaseName:postgres Statement:CREATE DATABASE IF NOT EXISTS postgres User:root}
I190212 07:34:23.843475 39528 server/server.go:1645  [n1] done ensuring all necessary migrations have run
I190212 07:34:23.843592 39528 server/server.go:1648  [n1] serving sql connections
I190212 07:34:23.870680 39789 sql/event_log.go:135  [n1] Event: "node_join", target: 1, info: {Descriptor:{NodeID:1 Address:127.0.0.1:44729 Attrs: Locality: ServerVersion:2.1-6 BuildTag:v2.2.0-alpha.20190211-59-g965525a StartedAt:1549956863221167191 LocalityAddress:[] XXX_NoUnkeyedLiteral:{} XXX_sizecache:0} ClusterID:0767d590-e3a2-4b92-839a-d55b5e214481 StartedAt:1549956863221167191 LastUp:1549956863221167191}
I190212 07:34:23.873853 39786 server/server_update.go:67  [n1] no need to upgrade, cluster already at the newest version
I190212 07:34:23.916926 39497 sql/event_log.go:135  [n1,client=127.0.0.1:52492,user=root] Event: "set_cluster_setting", target: 0, info: {SettingName:kv.rangefeed.enabled Value:true User:root}
I190212 07:34:23.945303 39497 sql/event_log.go:135  [n1,client=127.0.0.1:52492,user=root] Event: "set_cluster_setting", target: 0, info: {SettingName:kv.closed_timestamp.target_duration Value:00:00:01 User:root}
I190212 07:34:23.982669 39497 sql/event_log.go:135  [n1,client=127.0.0.1:52492,user=root] Event: "set_cluster_setting", target: 0, info: {SettingName:changefeed.experimental_poll_interval Value:00:00:00.01 User:root}
I190212 07:34:24.009996 39497 sql/event_log.go:135  [n1,client=127.0.0.1:52492,user=root] Event: "create_database", target: 52, info: {DatabaseName:d Statement:CREATE DATABASE d User:root}
I190212 07:34:24.045299 39497 sql/event_log.go:135  [n1,client=127.0.0.1:52492,user=root] Event: "set_cluster_setting", target: 0, info: {SettingName:changefeed.push.enabled Value:false User:root}
I190212 07:34:24.077778 39497 sql/event_log.go:135  [n1,client=127.0.0.1:52492,user=root] Event: "set_cluster_setting", target: 0, info: {SettingName:changefeed.experimental_poll_interval Value:00:00:00.01 User:root}
I190212 07:34:24.105689 39497 sql/event_log.go:135  [n1,client=127.0.0.1:52492,user=root] Event: "create_table", target: 53, info: {TableName:d.public.foo Statement:CREATE TABLE foo (a INT8 PRIMARY KEY, b STRING) User:root}
I190212 07:34:24.120701 39912 storage/replica_command.go:245  [n1,split,s1,r20/1:/{Table/24-Max}] initiating a split of this range at key /Table/53 [r21] (zone config)
I190212 07:34:24.474823 39528 util/stop/stopper.go:536  quiescing; tasks left:
1      [async] transport racer
1      [async] storage.IntentResolver: cleanup txn records
1      [async] intent_resolver_batcher
1      [async] closedts-rangefeed-subscriber
1      [async] changefeed-poller
1      [async] async-rollback
I190212 07:34:24.485080 39528 util/stop/stopper.go:536  quiescing; tasks left:
1      [async] transport racer
1      [async] storage.IntentResolver: cleanup txn records
1      [async] closedts-rangefeed-subscriber
W190212 07:34:24.485542 39896 storage/intentresolver/intent_resolver.go:721  failed to gc transaction record: could not GC completed transaction anchored at /Local/Range/Table/24/RangeDescriptor: node unavailable; try another peer
I190212 07:34:24.485810 39528 util/stop/stopper.go:536  quiescing; tasks left:
1      [async] transport racer
1      [async] closedts-rangefeed-subscriber
I190212 07:34:24.486039 39528 util/stop/stopper.go:536  quiescing; tasks left:
1      [async] transport racer
I190212 07:34:24.570205 39659 kv/transport_race.go:113  transport race promotion: ran 12 iterations on up to 604 requests
    --- FAIL: TestAvroEncoder/poller (1.49s)
        <autogenerated>:1: expected
              foo: {"a":{"long":3}}->{"after":{"foo":{"a":{"long":3},"b":{"string":"baz"}}},"updated":{"string":"1549956864460181344.0000000000"}}
            got
              foo: {"a":{"long":3}}->{"after":{"foo":{"a":{"long":3},"b":{"string":"baz"}}},"updated":{"string":"1549956864451566598.0000000001"}}

```

SHA: https://github.com/cockroachdb/cockroach/commits/10f8010fa5778e740c057905e2d7664b5fd5d647

Parameters:
```
TAGS=
GOFLAGS=-race
```

To repro, try:

```
# Don't forget to check out a clean suitable branch and experiment with the
# stress invocation until the desired results present themselves. For example,
# using stress instead of stressrace and passing the '-p' stressflag which
# controls concurrency.
./scripts/gceworker.sh start && ./scripts/gceworker.sh mosh
cd ~/go/src/github.com/cockroachdb/cockroach && \
stdbuf -oL -eL \
make stressrace TESTS=TestAvroEncoder PKG=github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl TESTTIMEOUT=5m STRESSFLAGS='-maxtime 20m -timeout 10m' 2>&1 | tee /tmp/stress.log
```

Failed test: https://teamcity.cockroachdb.com/viewLog.html?buildId=1135873&tab=buildLog

```
I190213 08:03:44.087778 39787 server/server.go:1648  [n1] serving sql connections
I190213 08:03:44.134447 40065 server/server_update.go:67  [n1] no need to upgrade, cluster already at the newest version
I190213 08:03:44.136415 40164 sql/event_log.go:135  [n1] Event: "node_join", target: 1, info: {Descriptor:{NodeID:1 Address:127.0.0.1:38655 Attrs: Locality: ServerVersion:2.1-6 BuildTag:v2.2.0-alpha.20190211-100-g10f8010 StartedAt:1550045023446407575 LocalityAddress:[] XXX_NoUnkeyedLiteral:{} XXX_sizecache:0} ClusterID:5fc734b7-1dcb-4b9e-9f10-20aa59bb23b2 StartedAt:1550045023446407575 LastUp:1550045023446407575}
I190213 08:03:44.169978 40137 sql/event_log.go:135  [n1,client=127.0.0.1:59328,user=root] Event: "set_cluster_setting", target: 0, info: {SettingName:kv.rangefeed.enabled Value:true User:root}
I190213 08:03:44.215123 40137 sql/event_log.go:135  [n1,client=127.0.0.1:59328,user=root] Event: "set_cluster_setting", target: 0, info: {SettingName:kv.closed_timestamp.target_duration Value:00:00:01 User:root}
I190213 08:03:44.251681 40137 sql/event_log.go:135  [n1,client=127.0.0.1:59328,user=root] Event: "set_cluster_setting", target: 0, info: {SettingName:changefeed.experimental_poll_interval Value:00:00:00.01 User:root}
I190213 08:03:44.309152 40137 sql/event_log.go:135  [n1,client=127.0.0.1:59328,user=root] Event: "create_database", target: 52, info: {DatabaseName:d Statement:CREATE DATABASE d User:root}
I190213 08:03:44.353301 40137 sql/event_log.go:135  [n1,client=127.0.0.1:59328,user=root] Event: "set_cluster_setting", target: 0, info: {SettingName:changefeed.push.enabled Value:false User:root}
I190213 08:03:44.392545 40137 sql/event_log.go:135  [n1,client=127.0.0.1:59328,user=root] Event: "set_cluster_setting", target: 0, info: {SettingName:changefeed.experimental_poll_interval Value:00:00:00.01 User:root}
I190213 08:03:44.437119 40137 sql/event_log.go:135  [n1,client=127.0.0.1:59328,user=root] Event: "create_table", target: 53, info: {TableName:d.public.foo Statement:CREATE TABLE foo (a INT8 PRIMARY KEY, b STRING) User:root}
I190213 08:03:44.459158 40242 storage/replica_command.go:245  [n1,split,s1,r20/1:/{Table/24-Max}] initiating a split of this range at key /Table/53 [r21] (zone config)
I190213 08:03:44.848236 39787 util/stop/stopper.go:547  quiescing; tasks left:
1      [async] transport racer
1      [async] storage.IntentResolver: cleanup txn records
1      [async] intent_resolver_batcher
1      [async] closedts-rangefeed-subscriber
1      [async] changefeed-poller
W190213 08:03:44.848313 40171 storage/intentresolver/intent_resolver.go:721  failed to gc transaction record: could not GC completed transaction anchored at /Local/Range/Table/24/RangeDescriptor: node unavailable; try another peer
I190213 08:03:44.848589 39787 util/stop/stopper.go:547  quiescing; tasks left:
1      [async] transport racer
1      [async] storage.IntentResolver: cleanup txn records
1      [async] intent_resolver_batcher
1      [async] closedts-rangefeed-subscriber
I190213 08:03:44.848796 39787 util/stop/stopper.go:547  quiescing; tasks left:
1      [async] transport racer
1      [async] intent_resolver_batcher
1      [async] closedts-rangefeed-subscriber
I190213 08:03:44.849114 39787 util/stop/stopper.go:547  quiescing; tasks left:
1      [async] transport racer
1      [async] closedts-rangefeed-subscriber
I190213 08:03:44.849332 39787 util/stop/stopper.go:547  quiescing; tasks left:
1      [async] transport racer
W190213 08:03:44.849458 40213 internal/client/txn.go:520  [n1,client=127.0.0.1:59336,user=root] failure aborting transaction: node unavailable; try another peer; abort caused by: context canceled
I190213 08:03:44.924444 40082 kv/transport_race.go:113  transport race promotion: ran 10 iterations on up to 623 requests
    --- FAIL: TestAvroEncoder/poller (1.65s)
        <autogenerated>:1: expected
              foo: {"a":{"long":3}}->{"after":{"foo":{"a":{"long":3},"b":{"string":"baz"}}},"updated":{"string":"1550045024829413917.0000000000"}}
            got
              foo: {"a":{"long":3}}->{"after":{"foo":{"a":{"long":3},"b":{"string":"baz"}}},"updated":{"string":"1550045024815525369.0000000001"}}

```

SHA: https://github.com/cockroachdb/cockroach/commits/22119c85d2e7aca50f21f61408ae55207679569d

Parameters:
```
TAGS=
GOFLAGS=-race
```

To repro, try:

```
# Don't forget to check out a clean suitable branch and experiment with the
# stress invocation until the desired results present themselves. For example,
# using stress instead of stressrace and passing the '-p' stressflag which
# controls concurrency.
./scripts/gceworker.sh start && ./scripts/gceworker.sh mosh
cd ~/go/src/github.com/cockroachdb/cockroach && \
stdbuf -oL -eL \
make stressrace TESTS=TestAvroEncoder PKG=github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl TESTTIMEOUT=5m STRESSFLAGS='-maxtime 20m -timeout 10m' 2>&1 | tee /tmp/stress.log
```

Failed test: https://teamcity.cockroachdb.com/viewLog.html?buildId=1140564&tab=buildLog

```
  github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTask()
      /go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:320 +0x14d
  github.com/cockroachdb/cockroach/pkg/jobs.(*Registry).resume()
      /go/src/github.com/cockroachdb/cockroach/pkg/jobs/registry.go:534 +0x307
  github.com/cockroachdb/cockroach/pkg/jobs.(*Registry).StartJob()
      /go/src/github.com/cockroachdb/cockroach/pkg/jobs/registry.go:224 +0x2f7
  github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl.changefeedPlanHook.func2()
      /go/src/github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl/changefeed_stmt.go:248 +0xf33
  github.com/cockroachdb/cockroach/pkg/sql.(*hookFnNode).startExec.func1()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/planhook.go:136 +0x108
==================
I190215 08:22:21.865895 40350 ccl/changefeedccl/changefeed_processors.go:550  [n1] job 426470739165216769 new resolved timestamp 1550218941.729762618,0 is behind by 136.037081ms
I190215 08:22:21.916404 40350 ccl/changefeedccl/changefeed_processors.go:550  [n1] job 426470739165216769 new resolved timestamp 1550218941.800727678,0 is behind by 115.590636ms
I190215 08:22:21.947317 40445 ccl/changefeedccl/changefeed_processors.go:550  [n1] job 426470740993507329 new resolved timestamp 1550218941.405653573,0 is behind by 541.530005ms
I190215 08:22:21.947618 40445 ccl/changefeedccl/changefeed_processors.go:562  [n1] job 426470740993507329 span /Table/53/{1-2} is behind by 541.530005ms
I190215 08:22:21.991576 40350 ccl/changefeedccl/changefeed_processors.go:550  [n1] job 426470739165216769 new resolved timestamp 1550218941.869488947,0 is behind by 121.938943ms
I190215 08:22:22.048809 40445 ccl/changefeedccl/changefeed_processors.go:550  [n1] job 426470740993507329 new resolved timestamp 1550218941.845130517,0 is behind by 203.58131ms
I190215 08:22:22.080475 40350 ccl/changefeedccl/changefeed_processors.go:550  [n1] job 426470739165216769 new resolved timestamp 1550218941.918866937,0 is behind by 161.517554ms
I190215 08:22:22.152191 40445 ccl/changefeedccl/changefeed_processors.go:550  [n1] job 426470740993507329 new resolved timestamp 1550218941.952422905,0 is behind by 199.685797ms
I190215 08:22:22.162285 40350 ccl/changefeedccl/changefeed_processors.go:550  [n1] job 426470739165216769 new resolved timestamp 1550218941.995299653,0 is behind by 166.888872ms
I190215 08:22:22.210719 40350 ccl/changefeedccl/changefeed_processors.go:550  [n1] job 426470739165216769 new resolved timestamp 1550218942.080909441,0 is behind by 129.686477ms
I190215 08:22:22.231518 40445 ccl/changefeedccl/changefeed_stmt.go:442  [n1] CHANGEFEED job 426470740993507329 returning with error: cannot update progress on canceled job (id 426470740993507329)
I190215 08:22:22.299970 40350 ccl/changefeedccl/changefeed_processors.go:550  [n1] job 426470739165216769 new resolved timestamp 1550218942.162657577,0 is behind by 137.184519ms
I190215 08:22:22.406194 39981 util/stop/stopper.go:547  quiescing; tasks left:
3      [async] storage.IntentResolver: cleanup txn records
1      [async] transport racer
1      [async] job-426470739165216769
1      [async] intent_resolver_batcher
1      [async] closedts-rangefeed-subscriber
1      [async] changefeed-poller
W190215 08:22:22.406794 40462 storage/intentresolver/intent_resolver.go:722  failed to gc transaction record: could not GC completed transaction anchored at /Table/SystemConfigSpan/Start: node unavailable; try another peer
I190215 08:22:22.410154 40350 ccl/changefeedccl/changefeed_stmt.go:442  [n1] CHANGEFEED job 426470739165216769 returning with error: cannot update progress on canceled job (id 426470739165216769)
I190215 08:22:22.411282 39981 util/stop/stopper.go:547  quiescing; tasks left:
1      [async] transport racer
1      [async] closedts-rangefeed-subscriber
I190215 08:22:22.415881 39981 util/stop/stopper.go:547  quiescing; tasks left:
1      [async] transport racer
I190215 08:22:22.564731 40242 kv/transport_race.go:113  transport race promotion: ran 35 iterations on up to 1338 requests
    --- FAIL: TestAvroEncoder/enterprise (2.84s)
        testing.go:771: race detected during execution of test

```

This reproduces locally under stressrace pretty easily. I turned on verbose changefeed logging and found something odd. I assumed the bug was in the avro encoder stuff reusing some memory that it shouldn't be but I _think_ I'm seeing a kv get written with a different timestamp than the one returned by `cluster_logical_timestamp()`.

The test does
```go
		var ts2 string
		require.NoError(t, crdb.ExecuteTx(ctx, db, nil /* txopts */, func(tx *gosql.Tx) error {
			return tx.QueryRow(
				`INSERT INTO foo VALUES (3, 'baz') RETURNING cluster_logical_timestamp()`,
			).Scan(&ts2)
		}))
		assertPayloadsAvro(t, reg, fooUpdated, []string{
			`foo: {"a":{"long":3}}->{"after":{"foo":{"a":{"long":3},"b":{"string":"baz"}}},` +
				`"updated":{"string":"` + ts2 + `"}}`,
		})
```

With verbose changefeed logging on (`--vmodule=changefeed=3`), I'm seeing this
```
I190226 19:28:51.117848 549 ccl/changefeedccl/changefeed.go:139  [n1,client=127.0.0.1:56320,user=root] changed key /Table/53/1/3/0 1551209331.061806000,1
...
--- FAIL: TestAvroEncoder (2.91s)
    --- FAIL: TestAvroEncoder/poller (2.91s)
        <autogenerated>:1: expected
              foo: {"a":{"long":3}}->{"after":{"foo":{"a":{"long":3},"b":{"string":"baz"}}},"updated":{"string":"1551209331098368000.0000000000"}}
            got
              foo: {"a":{"long":3}}->{"after":{"foo":{"a":{"long":3},"b":{"string":"baz"}}},"updated":{"string":"1551209331061806000.0000000001"}}
```

The "expected" in the error is ts2, which comes back through sql from `RETURNING cluster_logical_timestamp()` on the insert. The "got" matches the timestamp we received from the changed kv, which comes right out of (in the case of this "poller" subtest" the ExportRequest).

--- a/pkg/ccl/storageccl/engineccl/mvcc.go
+++ b/pkg/ccl/storageccl/engineccl/mvcc.go
@@ -81,6 +81,17 @@ func NewMVCCIncrementalIterator(e engine.Reader, opts IterOptions) *MVCCIncremen
 	return &MVCCIncrementalIterator{
 		e:          e,
 		upperBound: opts.UpperBound,
+		// It is necessary for correctness that sanityIter be created before
+		// iter. This is because the provided Reader may not be a consistent
+		// snapshot, so the two could end up observing different information.
+		// The hack around sanityCheckMetadataKey only works properly if all
+		// possible discrepancies between the two iterators lead to intents
+		// and values falling outside of the timestamp range **from iter's
+		// perspective**. This allows us to simply ignore discrepancies that
+		// we notice in advance(). See #34819.
+		sanityIter: e.NewIterator(engine.IterOptions{
+			UpperBound: opts.UpperBound,
+		}),
 		iter: e.NewIterator(engine.IterOptions{
 			// The call to startTime.Next() converts our exclusive start bound into
 			// the inclusive start bound that MinTimestampHint expects. This is
@@ -214,11 +225,6 @@ func (i *MVCCIncrementalIterator) advance() {
 // sees that exact key. Otherwise, it returns false. It's used in the workaround
 // in `advance` for a time-bound iterator bug.
 func (i *MVCCIncrementalIterator) sanityCheckMetadataKey() ([]byte, bool, error) {
-	if i.sanityIter == nil {
-		// The common case is that we'll won't need the sanityIter for a given
-		// MVCCIncrementalIterator, so create it lazily.
-		i.sanityIter = i.e.NewIterator(engine.IterOptions{UpperBound: i.upperBound})
-	}
 	unsafeKey := i.iter.UnsafeKey()
 	i.sanityIter.Seek(unsafeKey)
 	if ok, err := i.sanityIter.Valid(); err != nil {

--- a/pkg/ccl/storageccl/engineccl/mvcc_test.go
+++ b/pkg/ccl/storageccl/engineccl/mvcc_test.go
@@ -24,7 +24,9 @@ import (
 	"github.com/cockroachdb/cockroach/pkg/util/hlc"
 	"github.com/cockroachdb/cockroach/pkg/util/leaktest"
 	"github.com/cockroachdb/cockroach/pkg/util/uuid"
+	"github.com/pkg/errors"
 	"github.com/stretchr/testify/require"
+	"golang.org/x/sync/errgroup"
 )
 
 func iterateExpectErr(
@@ -486,6 +488,28 @@ func TestMVCCIncrementalIteratorIntentStraddlesSStables(t *testing.T) {
 	}
 }
 
+func slurpKVsInTimeRange(
+	e engine.Reader, prefix roachpb.Key, startTime, endTime hlc.Timestamp,
+) ([]engine.MVCCKeyValue, error) {
+	endKey := prefix.PrefixEnd()
+	iter := NewMVCCIncrementalIterator(e, IterOptions{
+		StartTime:  startTime,
+		EndTime:    endTime,
+		UpperBound: endKey,
+	})
+	defer iter.Close()
+	var kvs []engine.MVCCKeyValue
+	for iter.Seek(engine.MakeMVCCMetadataKey(prefix)); ; iter.Next() {
+		if ok, err := iter.Valid(); err != nil {
+			return nil, err
+		} else if !ok || iter.UnsafeKey().Key.Compare(endKey) >= 0 {
+			break
+		}
+		kvs = append(kvs, engine.MVCCKeyValue{Key: iter.Key(), Value: iter.Value()})
+	}
+	return kvs, nil
+}
+
 // TestMVCCIncrementalIteratorIntentDeletion checks a workaround in
 // MVCCIncrementalIterator for a bug in time-bound iterators, where an intent
 // has been deleted, but the time-bound iterator doesn't see the deletion.
@@ -509,27 +533,6 @@ func TestMVCCIncrementalIteratorIntentDeletion(t *testing.T) {
 			Txn:    txn.TxnMeta,
 			Status: roachpb.COMMITTED,
 		}
-	}
-	slurpKVs := func(
-		e engine.Reader, prefix roachpb.Key, startTime, endTime hlc.Timestamp,
-	) ([]engine.MVCCKeyValue, error) {
-		endKey := prefix.PrefixEnd()
-		iter := NewMVCCIncrementalIterator(e, IterOptions{
-			StartTime:  startTime,
-			EndTime:    endTime,
-			UpperBound: endKey,
-		})
-		defer iter.Close()
-		var kvs []engine.MVCCKeyValue
-		for iter.Seek(engine.MakeMVCCMetadataKey(prefix)); ; iter.Next() {
-			if ok, err := iter.Valid(); err != nil {
-				return nil, err
-			} else if !ok || iter.UnsafeKey().Key.Compare(endKey) >= 0 {
-				break
-			}
-			kvs = append(kvs, engine.MVCCKeyValue{Key: iter.Key(), Value: iter.Value()})
-		}
-		return kvs, nil
 	}
 
 	ctx := context.Background()
@@ -592,32 +595,102 @@ func TestMVCCIncrementalIteratorIntentDeletion(t *testing.T) {
 
 	// The kA ts1 intent has been resolved. There's now a new intent on kA, but
 	// the timestamp (ts3) is too new so it should be ignored.
-	kvs, err := slurpKVs(db, kA, ts0, ts1)
+	kvs, err := slurpKVsInTimeRange(db, kA, ts0, ts1)
 	require.NoError(t, err)
 	require.Equal(t, []engine.MVCCKeyValue{
 		{Key: engine.MVCCKey{Key: kA, Timestamp: ts1}, Value: vA1.RawBytes},
 	}, kvs)
 	// kA has a value at ts2. Again the intent is too new (ts3), so ignore.
-	kvs, err = slurpKVs(db, kA, ts0, ts2)
+	kvs, err = slurpKVsInTimeRange(db, kA, ts0, ts2)
 	require.NoError(t, err)
 	require.Equal(t, []engine.MVCCKeyValue{
 		{Key: engine.MVCCKey{Key: kA, Timestamp: ts2}, Value: vA2.RawBytes},
 		{Key: engine.MVCCKey{Key: kA, Timestamp: ts1}, Value: vA1.RawBytes},
 	}, kvs)
 	// At ts3, we should see the new intent
-	_, err = slurpKVs(db, kA, ts0, ts3)
+	_, err = slurpKVsInTimeRange(db, kA, ts0, ts3)
 	require.EqualError(t, err, `conflicting intents on "kA"`)
 
 	// Similar to the kA ts1 check, but there is no newer intent. We expect to
 	// pick up the intent deletion and it should cancel out the intent, leaving
 	// only the value at ts1.
-	kvs, err = slurpKVs(db, kB, ts0, ts1)
+	kvs, err = slurpKVsInTimeRange(db, kB, ts0, ts1)
 	require.NoError(t, err)
 	require.Equal(t, []engine.MVCCKeyValue{
 		{Key: engine.MVCCKey{Key: kB, Timestamp: ts1}, Value: vB1.RawBytes},
 	}, kvs)
 
 	// Sanity check that we see the still unresolved intent for kC ts1.
-	_, err = slurpKVs(db, kC, ts0, ts1)
+	_, err = slurpKVsInTimeRange(db, kC, ts0, ts1)
 	require.EqualError(t, err, `conflicting intents on "kC"`)
+}
+
+// TestMVCCIncrementalIteratorIntentRewrittenConcurrently verifies that the
+// workaround in MVCCIncrementalIterator to double-check for deleted intents
+// properly handles cases where an intent originally in a time-bound iterator's
+// time range is rewritten at a timestamp outside of its time range.
+func TestMVCCIncrementalIteratorIntentRewrittenConcurrently(t *testing.T) {
+	defer leaktest.AfterTest(t)()
+
+	// Create a DB containing a single intent.
+	ctx := context.Background()
+	db := engine.NewInMem(roachpb.Attributes{}, 10<<20 /* 10 MB */)
+	defer db.Close()
+
+	kA := roachpb.Key("kA")
+	vA1 := roachpb.MakeValueFromString("vA1")
+	vA2 := roachpb.MakeValueFromString("vA2")
+	ts0 := hlc.Timestamp{WallTime: 0}
+	ts1 := hlc.Timestamp{WallTime: 1}
+	ts2 := hlc.Timestamp{WallTime: 2}
+	ts3 := hlc.Timestamp{WallTime: 3}
+	txn := &roachpb.Transaction{
+		TxnMeta: enginepb.TxnMeta{
+			Key:       roachpb.Key("b"),
+			ID:        uuid.MakeV4(),
+			Epoch:     1,
+			Timestamp: ts1,
+			Sequence:  1,
+		},
+		OrigTimestamp: ts1,
+	}
+	if err := engine.MVCCPut(ctx, db, nil, kA, ts1, vA1, txn); err != nil {
+		t.Fatal(err)
+	}
+
+	// Concurrently iterate over the intent using a time-bound iterator and move
+	// the intent out of the time-bound iterator's time range by writing to it
+	// again at a higher timestamp.
+	g, _ := errgroup.WithContext(ctx)
+	g.Go(func() error {
+		// Re-write the intent with a higher timestamp.
+		txn.Timestamp = ts3
+		txn.Sequence = 2
+		return engine.MVCCPut(ctx, db, nil, kA, ts1, vA2, txn)
+	})
+	g.Go(func() error {
+		// Iterate with a time range that includes the initial intent but does
+		// not include the new intent.
+		kvs, err := slurpKVsInTimeRange(db, kA, ts0, ts2)
+
+		// There are two permissible outcomes from the scan. If the iteration
+		// wins the race with the put that moves the intent then it should
+		// observe the intent and return a write intent error. If the iteration
+		// loses the race with the put that moves the intent then it should
+		// observe and return nothing because there will be no committed or
+		// provisional keys in its time range.
+		if err != nil {
+			if !testutils.IsError(err, `conflicting intents on "kA"`) {
+				return err
+			}
+		} else {
+			if len(kvs) != 0 {
+				return errors.Errorf(`unexpected kvs: %v`, kvs)
+			}
+		}
+		return nil
+	})
+	if err := g.Wait(); err != nil {
+		t.Fatal(err)
+	}
 }

